File: IntersectionofTwoLinkedLists.cpp
#include<iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        while(headA){
            ListNode *p = headB;
            while(p && p!=headA){
                p = p->next;
            }
            if(p==headA) return p;
            headA = headA->next;
        }
        
        return NULL;
    }
};

int main(){
    ListNode *a1 = new ListNode(1);
    ListNode *a2 = new ListNode(2);
    
    ListNode *b1 = new ListNode(3);
    ListNode *b2 = new ListNode(4);
    ListNode *b3 = new ListNode(5);
    
    ListNode *c1 = new ListNode(6);
    ListNode *c2 = new ListNode(7);
    ListNode *c3 = new ListNode(8);
    
    a1->next = a2;
    a2->next = c1;
    c1->next = c2;
    c2->next = c3;
    b1->next = b2;
    b2->next = b3;
    b3->next = c1;
    
    Solution *solution = new Solution();
    cout<<solution->getIntersectionNode(a1, b1)->val<<endl;
    
    return 0;
    
}

File: IntersectionofTwoLinkedLists.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None
        
class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        a = headA
        b = headB
        while a != b:
            a = headB if not a else a.next
            b = headA if not b else b.next
        return a
        
if __name__=='__main__':
    a1 = ListNode(1)
    a2 = ListNode(2)
    
    b1 = ListNode(3)
    b2 = ListNode(4)
    b3 = ListNode(5)
    
    c1 = ListNode(6)
    c2 = ListNode(7)
    c3 = ListNode(8)
    
    a1.next = a2
    a2.next = c1
    c1.next = c2
    c2.next = c3
    b1.next = b2
    b2.next = b3
    b3.next = c1
    
    solution = Solution()
    print(solution.getIntersectionNode(a1, b1).val)
